File: last-round/40.组合总和-ii.js
/*
 * @lc app=leetcode.cn id=40 lang=javascript
 *
 * [40] 组合总和 II
 *
 * https://leetcode-cn.com/problems/combination-sum-ii/description/
 *
 * algorithms
 * Medium (63.63%)
 * Likes:    579
 * Dislikes: 0
 * Total Accepted:    158.1K
 * Total Submissions: 248.4K
 * Testcase Example:  '[10,1,2,7,6,1,5]\n8'
 *
 * 给定一个数组 candidates 和一个目标数 target ，找出 candidates 中所有可以使数字和为 target 的组合。
 * 
 * candidates 中的每个数字在每个组合中只能使用一次。
 * 
 * 说明：
 * 
 * 
 * 所有数字（包括目标数）都是正整数。
 * 解集不能包含重复的组合。 
 * 
 * 
 * 示例 1:
 * 
 * 输入: candidates = [10,1,2,7,6,1,5], target = 8,
 * 所求解集为:
 * [
 * ⁠ [1, 7],
 * ⁠ [1, 2, 5],
 * ⁠ [2, 6],
 * ⁠ [1, 1, 6]
 * ]
 * 
 * 
 * 示例 2:
 * 
 * 输入: candidates = [2,5,2,1,2], target = 5,
 * 所求解集为:
 * [
 * [1,2,2],
 * [5]
 * ]
 * 
 */

// @lc code=start
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
  const result = []

  const map = {}
  const candidatesSet = [...new Set(candidates)].sort()
  for (const c of candidates) map[c] = (map[c] || 0) + 1

  const dfs = (index, currentCombination, target) => {
    if (target === 0) result.push(currentCombination)
    if (index >= candidatesSet.length) return
    if (target <= 0) return

    const currentNum = candidatesSet[index]
    const maxSelect = map[currentNum]

    // 选 N 个
    for (let i = 0; i <= maxSelect; i++) {
      const arr = new Array(i).fill(currentNum)
      dfs(index + 1, [...currentCombination, ...arr], target - i * currentNum)
    }
  }
  dfs(0, [], target)

  return result
};
// @lc code=end

File: last-round/40.组合总和-ii.py
#
# @lc app=leetcode.cn id=40 lang=python3
#
# [40] 组合总和 II
#
# https://leetcode-cn.com/problems/combination-sum-ii/description/
#
# algorithms
# Medium (61.73%)
# Likes:    752
# Dislikes: 0
# Total Accepted:    218.2K
# Total Submissions: 353.5K
# Testcase Example:  '[10,1,2,7,6,1,5]\n8'
#
# 给定一个数组 candidates 和一个目标数 target ，找出 candidates 中所有可以使数字和为 target 的组合。
# 
# candidates 中的每个数字在每个组合中只能使用一次。
# 
# 注意：解集不能包含重复的组合。 
# 
# 
# 
# 示例 1:
# 
# 
# 输入: candidates = [10,1,2,7,6,1,5], target = 8,
# 输出:
# [
# [1,1,6],
# [1,2,5],
# [1,7],
# [2,6]
# ]
# 
# 示例 2:
# 
# 
# 输入: candidates = [2,5,2,1,2], target = 5,
# 输出:
# [
# [1,2,2],
# [5]
# ]
# 
# 
# 
# 提示:
# 
# 
# 1 
# 1 
# 1 
# 
# 
#

# @lc code=start
from collections import Counter


class Solution:
    def combinationSum2(self, candidates: list[int], target: int) -> list[list[int]]:
        result = []
        c = Counter(candidates)
        nums = list(c.keys())
        len_nums = len(nums)

        candidates.sort()

        def helper(prev_list: list[int], prev_index: int, prev_sum: int):
            if prev_sum == target:
                result.append(prev_list + [])
                return
            if prev_sum >= target or prev_index == len_nums - 1:
                return

            next_num = nums[prev_index + 1]
            iteration_count = 0

            for _ in range(c[next_num]):
                iteration_count += 1
                prev_list.append(next_num)
                prev_sum += next_num
                if prev_sum > target:
                    break
                helper(prev_list, prev_index + 1, prev_sum)

            for _ in range(iteration_count):
                prev_sum -= next_num
                prev_list.pop()
            helper(prev_list, prev_index + 1, prev_sum)

        helper([], -1, 0)
        return result
# @lc code=end

s = Solution()
print(s.combinationSum2([10,1,2,7,6,1,5], 8))
print(s.combinationSum2([2,5,2,1,2], 5))
